### DISCUSSION TITLE
Idea: Sort the macro context menu by macro name

### DISCUSSION THREAD

**Discussion by samaguire** [TabularEditor/TabularEditor3#1090] [Ideas]
As per the title. Can we get the context menu sorted by name with folders floated to the top?

![image](https://user-images.githubusercontent.com/62320770/155394550-1075f533-51e4-406d-a7c2-d5263a090214.png)

**Comment by otykier** [comment]
I think this should be an option. There could be good reasons for having macros and their folders appear in a specific, non-alphabetical order. Today, macros in the context menu are added in the order they appear in the `%localappdata%\TabularEditor3\MacroActions.json` file - this goes for their folders as well. So as a workaround, you could rearrange the items in the JSON array, to achieve the order you need.

**Comment by samaguire** [comment]
An option works. 👍 I’ll rearrange the JSON in the meantime. 😊


Regards
Stephen

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by samaguire** [comment]
> I think this should be an option. There could be good reasons for having macros and their folders appear in a specific, non-alphabetical order. Today, macros in the context menu are added in the order they appear in the `%localappdata%\TabularEditor3\MacroActions.json` file - this goes for their folders as well. So as a workaround, you could rearrange the items in the JSON array, to achieve the order you need.

For anyone you wants a quick & dirty sort:

``` csharp
using System.IO;
using Newtonsoft.Json.Linq;

var jsonFile = Environment.GetEnvironmentVariable("LocalAppData") + @"\TabularEditor3\MacroActions.json";
var json = Newtonsoft.Json.Linq.JObject.Parse(File.ReadAllText(jsonFile));
json["Actions"] = new JArray(json["Actions"].OrderBy(i => i["Name"]));
File.WriteAllText(jsonFile, JsonConvert.SerializeObject(json, Newtonsoft.Json.Formatting.Indented), System.Text.Encoding.UTF8);
ScriptHelper.Warning("Script finished. Please restart Tabular Editor 3.");
```
